site/viteUtils.tsx

```diff
@@ -0,0 +1,147 @@
+
+interface Assets {
+    forHeader: JSX.Element[]
+    forFooter: JSX.Element[]
+}
+
+// in dev: we need to load several vite core scripts and plugins; other than that we only need to load the entry point, and vite will take care of the rest.
+const devAssets = (entry: string): Assets => {
+    return {
+        forHeader: [googleFontsStyles],
+        forFooter: [
+            polyfillScript,
+            <script
+                key="vite-react-preamble" // https://vitejs.dev/guide/backend-integration.html
+                type="module"
+                dangerouslySetInnerHTML={{
+                    __html: `import RefreshRuntime from '${VITE_DEV_URL}/@react-refresh'
+  RefreshRuntime.injectIntoGlobalHook(window)
+  window.$RefreshReg$ = () => {}
+  window.$RefreshSig$ = () => (type) => type
+  window.__vite_plugin_react_preamble_installed__ = true`,
+                }}
+            />,
+            <script
+                key="vite-plugin-checker"
+                type="module"
+                src={`${VITE_DEV_URL}/@vite-plugin-checker-runtime-entry`}
+            />,
+            <script
+                key="vite-client"
+                type="module"
+                src={`${VITE_DEV_URL}/@vite/client`}
+            />,
+            <script
+                key={entry}
+                type="module"
+                src={`${VITE_DEV_URL}/${entry}`}
+            />,
+        ],
+    }
+}
+
+// Goes through the manifest.json file that vite creates, finds all the assets that are required for the entry point,
+// and creates the appropriate <link> and <script> tags for them.
+const createTagsForManifestEntry = (
+    manifest: Manifest,
+    entry: string,
+    assetBaseUrl: string
+): Assets => {
+    const createTags = (entry: string): JSX.Element[] => {
+        const manifestEntry =
+            Object.values(manifest).find((e) => e.file === entry) ??
+            manifest[entry]
+        let assets = [] as JSX.Element[]
+
+        if (!manifestEntry)
+            throw new Error(`Could not find manifest entry for ${entry}`)
+
+        const assetUrl = `${assetBaseUrl}${manifestEntry.file}`
+
+        if (entry.endsWith(".css")) {
+            assets = [
+                ...assets,
+                <link key={entry} rel="stylesheet" href={assetUrl} />,
+            ]
+        } else if (entry.match(/\.[cm]?(js|jsx|ts|tsx)$/)) {
+            // explicitly reference the entry; preload it and its dependencies
+            if (manifestEntry.isEntry) {
+                assets = [
+                    ...assets,
+                    <script key={entry} type="module" src={assetUrl} />,
+                ]
+            }
+
+            assets = [
+                ...assets,
+                <link
+                    key={`${entry}-preload`}
+                    rel="modulepreload" // see https://developer.chrome.com/blog/modulepreload/
+                    href={assetUrl}
+                />,
```

Comment:
Sadly, [Firefox doesn't support `modulepreload` at this point](https://caniuse.com/link-rel-modulepreload), and Safari support is in technical preview right now.
However, i[t looks like Firefox is actively working on the implementation](https://bugzilla.mozilla.org/show_bug.cgi?id=1425310), and like it will ship in the coming releases.

Until then, the `modulepreload` directive will not be taken into account, and pageloads are a bit slower sadly (because `common.mjs` can only be requested after `owid.mjs`, not in parallel).